pkg/cdi/registry.go

```diff
@@ -0,0 +1,119 @@
+	return GetRegistry().GetSpecDirectories()
+}
+
+// ListVendors lists all vendors known to the registry.
+func ListVendors() []string {
+	return GetRegistry().ListVendors()
+}
+
+// GetVendorSpecs returns all CDI Specs for a vendor.
+func GetVendorSpecs(vendor string) []*Spec {
+	return GetRegistry().GetVendorSpecs(vendor)
+}
+
+// ListDevices lists all qualified devices (matching an optional filter).
+func ListDevices(f *Filter) ([]string, error) {
+	return GetRegistry().ListDevices(f)
+}
+
+// MatchDevices lists all qualified devices matching a glob pattern.
+func MatchDevices(pattern string) ([]string, error) {
+	return GetRegistry().MatchDevices(pattern)
+}
+
+// HasDevice checks if the registry contains a qualified device.
+func HasDevice(device string) bool {
+	return GetRegistry().HasDevice(device)
+}
+
+// GetDevice returns the requested device from the registry.
+func GetDevice(device string) *Device {
+	return GetRegistry().GetDevice(device)
+}
+
+// HasErrors returns true if the registry population encountered any errors.
+func HasErrors() bool {
+	return GetRegistry().HasErrors()
+}
+
+// Errors returns any errors encountered during registry population.
+func Errors() []error {
+	errors := GetRegistry().Errors()
+	return errors
+}
+
+// InjectDevices injects the given qualified devices to an OCI Spec.
+// It returns unresolved devices and an error if injection fails.
+func InjectDevices(ociSpec *oci.Spec, devices []string) ([]string, error) {
+	return GetRegistry().InjectDevices(ociSpec, devices)
+}
+
+// ResolveDevices resolves any qualified devices present in an OCI Spec.
+// It returns the resolved qualified devices and an error if resolving fails.
+func ResolveDevices(ociSpec *oci.Spec) ([]string, error) {
+	return GetRegistry().ResolveDevices(ociSpec)
+}
+
+// Refresh reloads CDI Specs to the registry.
+func Refresh() error {
+	return GetRegistry().Refresh()
+}
+
+// EnableRefresh enables automatic reloading of updated Specs.
+func EnableRefresh() bool {
+	return GetRegistry().EnableRefresh()
+}
+
+// DisableRefresh disables automatic reloading of updated Specs.
+func DisableRefresh() bool {
+	return GetRegistry().DisableRefresh()
+}
+
+// BlockRefresh temporarily blocks automatic registry refresh.
+func BlockRefresh() {
+	GetRegistry().BlockRefresh()
+}
+
+// UnblockRefresh removes temporary blocking of automatic registry refresh.
+func UnblockRefresh() {
+	GetRegistry().UnblockRefresh()
+}
+
+// SetOptions applies the given options to the registry.
+func SetOptions(options ...RegistryOption) error {
+	return GetRegistry().SetOptions(options...)
+}
+
+// ScanSpecFiles scans the CDI Spec directories, reads each Spec file
+// and calls the scan function for each. ScanSpecFile does not modify
+// the registry itself. The scanned Specs are *NOT* loaded there.
+func ScanSpecFiles(fn ScanSpecsFunc) error {
+	return GetRegistry().ScanSpecFiles(fn)
+}
+
+// GetRegistry returns the CDI registry.
+func GetRegistry() *Registry {
```

Comment:
There are a few reasons. Some are merely a matter of taste/habit, others have real design/usability impact.

The habitual reason: I almost never use package scope as pseudo-object. Even if I know that semantically I'm writing a singleton I rather roll a real type for it which I either instantiate from `init()` or on-demand from `sync.One`-protected functions, like here.

About the mirrored functions... it is a fairly common practice (logrus, other loggers, standard flag package, etc.).

Regarding the `cdi.Foo()` vs. `cdi.GetRegistry().Foo()` comment, the answer is partly the same as above. But with the latest update I also tried to address Ed's complaint about the registry API having a lot of functions which might be useful in some cases but are definitely not needed for the vast majority usage. I used the 'package-level mirrored functions' to reduce the footprint of the API a user must expose himself to without loss of functionality. The rest is still available if you go the `cdi.GetRegistry()`-route.

Reply:
As I mentioned in my comment in #38 I think we have different types of clients and use cases that we are trying to address which leads to a more complex API at this stage. Does it make sense to separate these further, implementing the logic (e.g. the `Cache`) with additional functions that can be used when implementing utilities for "power" users, but returning a simple interface for normal interactions with container engines.

As an example, consider onboarding the `nvidia-container-runtime` to CDI as a thought exercise. Here we would check the `NVIDIA_VISIBLE_DEVICES` envvar for a fully-qualified CDI device ID, load the OCI spec from file, apply the required modifications, and write it back to disk before exec-ing to a low-level runtime such as `runc`:

```
import (
github.com/container-orchestrated-devices/container-device-interface/pkg/cdi
oci ...
)
func modify(spec *oci.Spec, deviceIDs ...string) error {
    registry := cdi.GetRegistry()
    specs, err := registry.Get(deviceIDs)
    if err != nil {
        return err
    }
    for _, s := specs {
       err := s.ApplyContainerEdits(s)
       if err != nil {
           return err
       }
    }
}
```
My current understanding is that apart from more advanced use cases like resolving devices from device host paths instead of requiring fully-qualified device IDs, the flow would be similar for other clients. Does it then make sense to focus on this functionality for the first iteration and then to expand it from there.

I do see that this is common practice in loggers or flag packages for example, but would argue in this case the usage of the API is limited to a specific portion of code in the clients and as such this kind of global access may not be required. I would prefer a constructor an interface to create an instance that is referenced on the client side and expose the API through that interface. This allows us to extend the "types" of registries we support or allows advanced users to implement their own. For example, for once off use cases, there is no need to manage refreshes etc.

Reply:
> Does it make sense to separate these further, implementing the logic (e.g. the Cache) with additional functions that can be used when implementing utilities for "power" users, but returning a simple interface for normal interactions with container engines.

Okay, so would you be more content if
  - `Registry` would not be an type alias for `Cache` but it would be instead an explicitly defined `type Registry interface{...}`
  - having same subset of Cache functions which are now package-level functions in `registry.go`
  - and we'd have the getter signature as `func GetRegistry() Registry`
  - we'd either turn `type Cache struct{}` into a pair of `type cache struct{}`, `type Cache interface{}`, or
  - if we can split Cache up in a meaningful way to multiple interfaces, each self-contained subset of the current Cache and semantically orthogonal wrt. the rest, then we'd rather go with that

Because I could probably live with that.

That said, generally speaking I don't see the point of splitting up an interface just for the sake of having smaller interfaces, unless there is really an orthogonal set of functionality being abstracted and a mish-mash of other abstractions present which need to consume those smaller interfaces. Also, I don't see much point in rolling an interface if it is not expected to be multiple implementations of it. It usually does not bring any benefits, often quite the opposite. The only exception to this is when one deals with really complex functionality which then needs to be easily mocked for unit- and other testing at the consumer side of the interface(s). And in this case, currently I neither see the existence of or the need for a mish-mash of interfaces cross-using each other, nor the really complex functionality that would need to be mocked in CDI consumers for testing. But it is of course possible that I'm missing something.

I'm just trying to understand where the shoe pinches and why...

Reply:
With the proposed PR I'd like to avoid CDI consumers having to write this kind of code below 
```
...
func modify(spec *oci.Spec, deviceIDs ...string) error {
    registry := cdi.GetRegistry()
    specs, err := registry.Get(deviceIDs)
    if err != nil {
        return err
    }
    for _, s := specs {
       err := s.ApplyContainerEdits(s)
       if err != nil {
           return err
       }
    }
}
```
and instead get along with a mere
```
func modify(spec *oci.Spec, devices []string) error {
   return err cdi.InjectDevices(ociSpec, devices)
   // or a cdi.GetRegistry().InjectDevices(ociSpec, devices) 
}
```
This will be the single most prevalent use case for CDI. This is what you need to do if you handle CDI on the client side and this is what you need to do if you handle it in the runtime. IMO, therefore it should possible to do it with an almost oneliner by calling a convenience function of the CDI package.